File: server/db/dataset.ts
import type { DatasetPayload, DatasetRow, DatasetRowValue, DatasetWithRows } from "~/types/dataset";
import { createSearchHash, db, decryptData, encryptData } from "$/db";
import { toNonSharedBytes } from "$/lib/utils";

const DATASET_ENCRYPTION_KEY = toNonSharedBytes(process.env.DATASET_ENCRYPTION_KEY, 32, false);
const key = await crypto.subtle.importKey("raw", DATASET_ENCRYPTION_KEY, "AES-GCM", false, [
    "encrypt",
    "decrypt",
]);

// PREDEFINED PREPARED STATEMENTS
const ADD_DATASET_QUERY = db.prepare<{ id: string }, [Uint8Array, Uint8Array]>(
    "INSERT INTO datasets (creator_user_hash, payload) VALUES (?, ?) RETURNING id",
);
const GET_ALL_DATASETS = db.query<{ id: string; payload: Uint8Array }, [Uint8Array]>(
    "SELECT id, payload FROM datasets WHERE creator_user_hash = ?",
);
const FIND_DATASET_BY_ID_QUERY = db.query<{ payload: Uint8Array }, [Uint8Array, string]>(
    "SELECT payload FROM datasets WHERE creator_user_hash = ? AND id = ?",
);
const UPDATE_DATASET_BY_ID_QUERY = db.prepare(
    "UPDATE datasets SET payload = ? WHERE creator_user_hash = ? AND id = ?",
);
const REMOVE_DATASET_BY_ID_QUERY = db.prepare(
    "DELETE FROM datasets WHERE creator_user_hash = ? AND id = ?",
);

const ADD_DATASET_ROW_QUERY = db.prepare(
    "INSERT INTO dataset_rows (dataset_id, key_hash, payload) VALUES ($dataset_id, $key_hash, $payload)",
);
const ADD_DATASET_ROWS_QUERY = db.transaction(
    (data: { $dataset_id: string; $key_hash: Uint8Array; $payload: Uint8Array }[]) => {
        for (const row of data) {
            ADD_DATASET_ROW_QUERY.run(row);
        }
        return data.length;
    },
);
const FIND_DATASET_ROWS_QUERY = db.query<{ payload: Uint8Array }, [string, Uint8Array]>(
    "SELECT payload FROM dataset_rows WHERE dataset_id = ? AND key_hash = ?",
);
const FIND_DATASET_ROWS_BY_DATASET_ID_QUERY = db.query<{ payload: Uint8Array }, [string]>(
    "SELECT payload FROM dataset_rows WHERE dataset_id = ?",
);
const FIND_DATASET_ROWS_BY_PROJECT_ID_QUERY_AND_KEY_HASH = db.query<
    { payload: Uint8Array },
    [string, Uint8Array]
>(
    "SELECT r.payload from dataset_rows r JOIN projects p ON r.dataset_id = p.dataset_id WHERE p.id = ? AND r.key_hash = ?",
);
const FIND_DATASET_ROWS_BY_PROJECT_ID_QUERY = db.query<{ payload: Uint8Array }, [string]>(
    "SELECT r.payload from dataset_rows r JOIN projects p ON r.dataset_id = p.dataset_id WHERE p.id = ?",
);
const REMOVE_DATASET_ROWS_QUERY = db.prepare(
    "DELETE FROM dataset_rows WHERE dataset_id = ? AND key_hash = ?",
);
const REMOVE_DATASET_ROWS_BY_DATASET_ID_QUERY = db.prepare(
    "DELETE FROM dataset_rows WHERE dataset_id = ?",
);

export async function addDataset(
    userHash: Uint8Array,
    dataset: DatasetPayload,
    rows?: DatasetRow[],
) {
    try {
        const payload = await encryptData(JSON.stringify(dataset), key);
        const row = ADD_DATASET_QUERY.get(userHash, payload);
        if (!row) return null;

        if (rows) await addDatasetRows(userHash, row.id, rows, dataset.key);

        return row.id;
    } catch {
        return null;
    }
}

export async function addDatasetRows(
    userHash: Uint8Array,
    datasetId: string,
    rows: DatasetRow[],
    datasetKey?: DatasetPayload["key"],
) {
    try {
        if (rows.length === 0) return 0;

        const _k = datasetKey || (await findDatasetById(userHash, datasetId))?.key;
        if (!_k) return 0;

        const data: Parameters<typeof ADD_DATASET_ROWS_QUERY>[0] = new Array(rows.length);
        for (let i = 0; i < rows.length; i++) {
            const row = rows[i];
            data[i] = {
                $dataset_id: datasetId,
                $key_hash: createSearchHash(row[_k]),
                $payload: await encryptData(JSON.stringify(row), key),
            };
        }

        return ADD_DATASET_ROWS_QUERY(data);
    } catch {
        return 0;
    }
}

export async function getAllDatasets(userHash: Uint8Array) {
    try {
        const rows = GET_ALL_DATASETS.all(userHash);
        return await Promise.all(
            rows.map(async (row) => ({
                id: row.id,
                ...(await decryptData<DatasetPayload>(row.payload, key)),
            })),
        );
    } catch {
        return [];
    }
}

export async function findDatasetById(userHash: Uint8Array, id: string, withRows = false) {
    try {
        const row = FIND_DATASET_BY_ID_QUERY.get(userHash, id);
        if (!row) return null;

        const res: DatasetWithRows | null = await decryptData<DatasetPayload>(row.payload, key);
        if (!res) return null;

        if (withRows) {
            const rows = await findDatasetRows(id, false);
            res.rows = rows;
        }

        return res;
    } catch {
        return null;
    }
}

export async function findDatasetRow(
    id: string, // If isProject is true, id is project id, else id is dataset id
    isProject: boolean,
    keyHash?: Uint8Array | DatasetRowValue,
) {
    try {
        const kh = typeof keyHash === "string" ? createSearchHash(keyHash) : keyHash;
        let payload: Uint8Array | null = null;
        if (kh) {
            if (isProject)
                payload =
                    FIND_DATASET_ROWS_BY_PROJECT_ID_QUERY_AND_KEY_HASH.get(id, kh)?.payload ?? null;
            else payload = FIND_DATASET_ROWS_QUERY.get(id, kh)?.payload ?? null;
        } else {
            if (isProject) payload = FIND_DATASET_ROWS_BY_PROJECT_ID_QUERY.get(id)?.payload ?? null;
            else payload = FIND_DATASET_ROWS_BY_DATASET_ID_QUERY.get(id)?.payload ?? null;
        }
        if (!payload) return null;
        return await decryptData<DatasetRow>(payload, key);
    } catch {
        return null;
    }
}
export async function findDatasetRows(
    id: string, // If isProject is true, id is project id, else id is dataset id
    isProject: boolean,
    keyHash?: Uint8Array | DatasetRowValue,
) {
    try {
        const kh = typeof keyHash === "string" ? createSearchHash(keyHash) : keyHash;
        let rows: { payload: Uint8Array }[] = [];
        if (kh) {
            if (isProject) rows = FIND_DATASET_ROWS_BY_PROJECT_ID_QUERY_AND_KEY_HASH.all(id, kh);
            else rows = FIND_DATASET_ROWS_QUERY.all(id, kh);
        } else {
            if (isProject) rows = FIND_DATASET_ROWS_BY_PROJECT_ID_QUERY.all(id);
            else rows = FIND_DATASET_ROWS_BY_DATASET_ID_QUERY.all(id);
        }
        return await Promise.all(rows.map((row) => decryptData<DatasetRow>(row.payload, key)));
    } catch {
        return [];
    }
}

export async function updateDataset(
    userHash: Uint8Array,
    id: string,
    datasetsPayload: Record<string, unknown>,
) {
    const prev = await findDatasetById(userHash, id);
    if (!prev) return 0;
    for (const k in prev) {
        if (!Object.hasOwn(prev, k)) continue;
        if (!Object.hasOwn(datasetsPayload, k)) continue;
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        prev[k as keyof typeof prev] = datasetsPayload[k] as any;
    }
    const payload = await encryptData(JSON.stringify(prev), key);
    const { changes } = UPDATE_DATASET_BY_ID_QUERY.run(payload, userHash, id);
    return changes;
}

export function removeDatasetById(userHash: Uint8Array, id: string) {
    return REMOVE_DATASET_BY_ID_QUERY.run(userHash, id).changes > 0;
}

export function removeDatasetRows(datasetId: string, keyHash?: Uint8Array) {
    if (keyHash) return REMOVE_DATASET_ROWS_QUERY.run(datasetId, keyHash).changes > 0;
    return REMOVE_DATASET_ROWS_BY_DATASET_ID_QUERY.run(datasetId).changes > 0;
}

File: server/db/dataset.go
package db

import (
	"crypto/cipher"
	"database/sql"
	"embed"
	"encoding/json"
	"os"
	"sync"

	"premark/lib"
	"premark/types"
)

var (
	datasetGCMInstance cipher.AEAD
	datasetGCMOnce     sync.Once
	datasetGCMErr      error
)

//go:embed sql/queries/datasets
var datasetsQueries embed.FS

//go:embed sql/queries/dataset_rows
var datasetRowsQueries embed.FS

func getDatasetGCM() (cipher.AEAD, error) {
	datasetGCMOnce.Do(func() {
		keyStr := os.Getenv("DATASET_ENCRYPTION_KEY")
		keyBytes := lib.ToNonSharedBytes(keyStr, 32, false)
		datasetGCMInstance, datasetGCMErr = lib.NewGCMHelper(keyBytes)
	})
	return datasetGCMInstance, datasetGCMErr
}

func AddDataset(userHash []byte, dataset types.DatasetPayload, rows []types.DatasetRow) (string, error) {
	gcm, err := getDatasetGCM()
	if err != nil {
		return "", err
	}

	payload, err := encryptJSON(dataset, gcm)
	if err != nil {
		return "", err
	}

	query, err := datasetsQueries.ReadFile("sql/queries/datasets/add.sql")
	if err != nil {
		return "", err
	}

	var id string
	err = DB.QueryRow(string(query), userHash, payload).Scan(&id)
	if err != nil {
		return "", err
	}

	if len(rows) > 0 {
		_, err = AddDatasetRows(userHash, id, rows, dataset.Key)
		if err != nil {
			return "", err
		}
	}

	return id, nil
}

func AddDatasetRows(userHash []byte, datasetId string, rows []types.DatasetRow, datasetKey ...string) (int, error) {
	if len(rows) == 0 {
		return 0, nil
	}

	var targetKey string
	if len(datasetKey) > 0 && datasetKey[0] != "" {
		targetKey = datasetKey[0]
	} else {
		prev, err := FindDatasetById(userHash, datasetId, false)
		if err != nil || prev == nil {
			return 0, err
		}
		targetKey = prev.Key
	}

	gcm, err := getDatasetGCM()
	if err != nil {
		return 0, err
	}

	tx, err := DB.Begin()
	if err != nil {
		return 0, err
	}
	defer tx.Rollback()

	query, err := datasetRowsQueries.ReadFile("sql/queries/dataset_rows/add.sql")
	if err != nil {
		return 0, err
	}
	stmt, err := tx.Prepare(string(query))
	if err != nil {
		return 0, err
	}
	defer stmt.Close()

	count := 0
	for _, row := range rows {
		kh, err := lib.CreateSearchHash(row[targetKey])
		if err != nil {
			return 0, err
		}

		payload, err := encryptJSON(row, gcm)
		if err != nil {
			return 0, err
		}

		if _, err := stmt.Exec(datasetId, kh, payload); err != nil {
			return 0, err
		}
		count++
	}

	return count, tx.Commit()
}

func GetAllDatasets(userHash []byte) ([]types.DatasetWithRows, error) {
	query, err := datasetsQueries.ReadFile("sql/queries/datasets/get_all.sql")
	if err != nil {
		return nil, err
	}

	rows, err := DB.Query(string(query), userHash)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	gcm, err := getDatasetGCM()
	if err != nil {
		return nil, err
	}

	var items []types.DatasetWithRows
	for rows.Next() {
		var id string
		var payload []byte
		if err := rows.Scan(&id, &payload); err != nil {
			return nil, err
		}

		ds, err := decryptJSON[types.DatasetPayload](payload, gcm)
		if err != nil {
			continue
		}
		items = append(items, types.DatasetWithRows{DatasetPayload: *ds, ID: id})
	}
	return items, nil
}

func FindDatasetById(userHash []byte, id string, withRows bool) (*types.DatasetWithRows, error) {
	query, err := datasetsQueries.ReadFile("sql/queries/datasets/find_by_id.sql")
	if err != nil {
		return nil, err
	}

	var payload []byte
	err = DB.QueryRow(string(query), userHash, id).Scan(&payload)
	if err == sql.ErrNoRows {
		return nil, nil
	} else if err != nil {
		return nil, err
	}

	gcm, err := getDatasetGCM()
	if err != nil {
		return nil, err
	}

	ds, err := decryptJSON[types.DatasetPayload](payload, gcm)
	if err != nil {
		return nil, err
	}

	res := &types.DatasetWithRows{DatasetPayload: *ds, ID: id}
	if withRows {
		dr, err := FindDatasetRows(id, false, nil)
		if err != nil {
			return nil, err
		}
		res.Rows = dr
	}
	return res, nil
}

func FindDatasetRow(id string, isProject bool, keyHash any) (types.DatasetRow, error) {
	res, err := FindDatasetRows(id, isProject, keyHash)
	if err != nil || len(res) == 0 {
		return nil, err
	}
	return res[0], nil
}

func FindDatasetRows(id string, isProject bool, keyHash any) ([]types.DatasetRow, error) {
	var kh []byte
	var err error

	if keyHash != nil {
		switch v := keyHash.(type) {
		case string:
			kh, err = lib.CreateSearchHash(v)
		case []byte:
			kh = v
		}
		if err != nil {
			return nil, err
		}
	}

	var query []byte
	var args []any
	if len(kh) > 0 {
		if isProject {
			query, err = datasetRowsQueries.ReadFile("sql/queries/dataset_rows/find_by_project_id_and_key_hash.sql")
			if err != nil {
				return nil, err
			}
		} else {
			query, err = datasetRowsQueries.ReadFile("sql/queries/dataset_rows/find_by_dataset_id_and_key_hash.sql")
			if err != nil {
				return nil, err
			}
		}
		args = append(args, id, kh)
	} else {
		if isProject {
			query, err = datasetRowsQueries.ReadFile("sql/queries/dataset_rows/find_by_project_id.sql")
			if err != nil {
				return nil, err
			}
		} else {
			query, err = datasetRowsQueries.ReadFile("sql/queries/dataset_rows/find_by_dataset_id.sql")
			if err != nil {
				return nil, err
			}
		}
		args = append(args, id)
	}

	rows, err := DB.Query(string(query), args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	gcm, err := getDatasetGCM()
	if err != nil {
		return nil, err
	}

	var list []types.DatasetRow
	for rows.Next() {
		var payload []byte
		if err := rows.Scan(&payload); err != nil {
			return nil, err
		}
		dr, err := decryptJSON[types.DatasetRow](payload, gcm)
		if err != nil {
			continue
		}
		list = append(list, *dr)
	}
	return list, nil
}

func UpdateDataset(userHash []byte, id string, datasetsPayload map[string]any) (int64, error) {
	prev, err := FindDatasetById(userHash, id, false)
	if err != nil || prev == nil {
		return 0, err
	}

	// Marshals through a generic structural map to update properties dynamically
	prevBytes, _ := json.Marshal(prev.DatasetPayload)
	var intermediate map[string]any
	_ = json.Unmarshal(prevBytes, &intermediate)

	for k, v := range datasetsPayload {
		if _, exists := intermediate[k]; exists {
			intermediate[k] = v
		}
	}

	var target types.DatasetPayload
	updatedBytes, _ := json.Marshal(intermediate)
	_ = json.Unmarshal(updatedBytes, &target)

	gcm, err := getDatasetGCM()
	if err != nil {
		return 0, err
	}

	payload, err := encryptJSON(target, gcm)
	if err != nil {
		return 0, err
	}

	query, err := datasetsQueries.ReadFile("sql/queries/datasets/update_by_id.sql")
	if err != nil {
		return 0, err
	}
	res, err := DB.Exec(string(query), payload, userHash, id)
	if err != nil {
		return 0, err
	}
	return res.RowsAffected()
}

func RemoveDatasetById(userHash []byte, id string) (bool, error) {
	query, err := datasetsQueries.ReadFile("sql/queries/datasets/delete_by_id.sql")
	if err != nil {
		return false, err
	}
	res, err := DB.Exec(string(query), userHash, id)
	if err != nil {
		return false, err
	}
	count, err := res.RowsAffected()
	return count > 0, err
}

func RemoveDatasetRows(datasetId string, keyHash ...[]byte) (bool, error) {
	var res sql.Result
	var err error
	var query []byte

	if len(keyHash) > 0 && len(keyHash[0]) > 0 {
		query, err = datasetRowsQueries.ReadFile("sql/queries/dataset_rows/delete_by_dataset_id_and_key_hash.sql")
		if err == nil {
			res, err = DB.Exec(string(query), datasetId, keyHash[0])
		}
	} else {
		query, err = datasetRowsQueries.ReadFile("sql/queries/dataset_rows/delete_by_dataset_id.sql")
		if err == nil {
			res, err = DB.Exec(string(query), datasetId)
		}
	}

	if err != nil {
		return false, err
	}
	count, err := res.RowsAffected()
	return count > 0, err
}
